core-util/v2/impl/FP1.hpp

```diff
@@ -0,0 +1,272 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __FUNCTIONPOINTER_IMPL_FP1_HPP__
+#define __FUNCTIONPOINTER_IMPL_FP1_HPP__
+#ifndef __FUNCTIONPOINTER_HPP__
+#error FP1.hpp is an internal file that should only be called from inside FunctionPointer.hpp
+#endif
+
+template <typename R, typename A1, size_t I>
+class FunctionPointer<R(A1), I> {
+public:
+    FunctionPointer(const FunctionPointer &fp) {
+        fp.const_ref().copy_to(_storage);
+    }
+    typedef R(*static_fp)(A1);
+    FunctionPointer(R (*fp)(A1)) {
+        new(_storage) impl::StaticPointer<R(A1), I>(fp);
+    }
+    template <typename C>
+    FunctionPointer(C *c, R (C::*member)(A1)) {
+        new(_storage) impl::MethodPointer<R(A1), C, I>(c, member);
+    }
+    template <typename F>
+    FunctionPointer(const F & f) {
+        new(_storage) impl::FunctorPointer<R(A1), F, I>(f);
+    }
+    operator bool() const {
+        return const_ref();
+    }
+    bool operator ==(const impl::FunctionPointerInterface<R(A1)> & rhs) const {
+        return const_ref() == rhs;
+    }
+    inline R operator ()(const A1&& a1) const {
```

Comment:
I don't think that you want to use a const rvalue references here, making this function a template function leverage perfect forwarding.